### PR TITLE
[C++ Worker]Add api get actor

### PR DESCRIPTION
**Pull request by qicosmos**
## Why are these changes needed?

Add API: GetActor.

## Checks

- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests

### CODE REVIEW

File: cpp/src/ray/runtime/task/native_task_submitter.cc [hunk 1]
```diff
@@ -83,5 +86,28 @@ ObjectID NativeTaskSubmitter::SubmitActorTask(InvocationSpec &invocation,
   return Submit(invocation, task_options);
 }
 
+std::string NativeTaskSubmitter::GetFullName(bool global, const std::string &name) {
+  if (name.empty()) {
+    return "";
+  }
+  return global ? name
+                : CoreWorkerProcess::GetCoreWorker().GetCurrentJobId().Hex() + "-" + name;
+}
```
Comment (kfstorm) [inline cpp/src/ray/runtime/task/native_task_submitter.cc:95]: Why write this method twice?

File: cpp/src/ray/runtime/task/native_task_submitter.cc [hunk 2]
```diff
@@ -83,5 +86,28 @@ ObjectID NativeTaskSubmitter::SubmitActorTask(InvocationSpec &invocation,
...
+  if (!pair.second.ok()) {
+    RAY_LOG(WARNING) << pair.second.message();
+    return ActorID::Nil();
+  }
+
+  std::string actor_id;
+  auto actor_handle = pair.first;
+  RAY_CHECK(actor_handle);
+  return actor_handle->GetActorID();
+}
```
Comment (kfstorm) [inline cpp/src/ray/runtime/task/native_task_submitter.cc:110]: Same here

File: cpp/src/ray/test/cluster/cluster_mode_test.cc [hunk 3]
```diff
@@ -240,6 +240,22 @@ TEST(RayClusterModeTest, ExceptionTest) {
...
+  auto named_actor_obj = actor.Task(&Counter::Plus1).Remote();
+  EXPECT_EQ(1, *named_actor_obj.Get());
+
+  auto named_actor_handle_optional = ray::GetActor<Counter>("named_actor");
+  EXPECT_TRUE(named_actor_handle_optional);
+  auto &named_actor_handle = *named_actor_handle_optional;
+  auto named_actor_obj1 = named_actor_handle.Task(&Counter::Plus1).Remote();
+  EXPECT_EQ(2, *named_actor_obj1.Get());
+  EXPECT_FALSE(ray::GetActor<Counter>("not_exist_actor"));
+}
```
Comment (kfstorm) [inline cpp/src/ray/test/cluster/cluster_mode_test.cc:257]: Why write this test case twice?

File: cpp/src/ray/runtime/task/native_task_submitter.cc (hunk 1 above)
Comment (qicosmos) [reply cpp/src/ray/runtime/task/native_task_submitter.cc:95]: GetFullName is a virtual function, one is for native mode, another is for local mode.

File: cpp/src/ray/runtime/task/native_task_submitter.cc (hunk 2 above)
Comment (qicosmos) [reply cpp/src/ray/runtime/task/native_task_submitter.cc:110]: GetActor is a virtual function, one is for native mode, another is for local mode.

File: cpp/src/ray/test/cluster/cluster_mode_test.cc (hunk 3 above)
Comment (qicosmos) [reply cpp/src/ray/test/cluster/cluster_mode_test.cc:257]: One is for local mode testing, another one is for cluster mode testing.

File: cpp/src/ray/runtime/task/native_task_submitter.cc (hunk 1 above)
Comment (kfstorm) [reply cpp/src/ray/runtime/task/native_task_submitter.cc:95]: Can we avoid duplicated code? Move the logic to the base class?

File: cpp/src/ray/test/cluster/cluster_mode_test.cc (hunk 3 above)
Comment (kfstorm) [reply cpp/src/ray/test/cluster/cluster_mode_test.cc:257]: Why not write the test once and add a switch flag to test it in both local and cluster mode?

File: cpp/src/ray/runtime/task/native_task_submitter.cc (hunk 1 above)
Comment (qicosmos) [reply cpp/src/ray/runtime/task/native_task_submitter.cc:95]: Not duplicated code, the implementation are different:
native mode:
CoreWorkerProcess::GetCoreWorker().GetCurrentJobId().Hex()
local mode:
local_mode_ray_tuntime_.GetCurrentJobID().Hex()

File: cpp/src/ray/test/cluster/cluster_mode_test.cc (hunk 3 above)
Comment (qicosmos) [reply cpp/src/ray/test/cluster/cluster_mode_test.cc:257]: Our current test cases are divided into two parts:
Local mode test: includes api_test.cc, ray_remote_test.cc
Cluster mode test: includes cluster_mode_test.cc
They have different singleton runtime after init, can't support both local mode and runtime mode at the same time now.

**Review by SongGuyang — COMMENTED** [review]
Can you update the document at the same time?

File: cpp/include/ray/api.h [hunk 4]
```diff
@@ -109,6 +109,16 @@ ray::internal::ActorCreator<F> Actor(F create_func);
 template <typename T>
 boost::optional<ActorHandle<T>> GetGlobalActor(const std::string &actor_name);
 
+/// Get a handle to a named actor of current job.
+/// Gets a handle to a named actor with the given name. The actor must have been created
+/// with name specified.
```
[the remark this answers is not in the document]
Comment (qicosmos) [reply cpp/include/ray/api.h:114]: The comment is a little similar with GetGlobalActor Api, it is copied from java api:
https://github.com/ray-project/ray/blob/master/java/api/src/main/java/io/ray/api/Ray.java#L162
https://github.com/ray-project/ray/blob/master/java/api/src/main/java/io/ray/api/Ray.java#L176

File: cpp/src/ray/runtime/task/native_task_submitter.cc [hunk 5]
```diff
@@ -83,5 +86,24 @@ ObjectID NativeTaskSubmitter::SubmitActorTask(InvocationSpec &invocation,
...
+ActorID NativeTaskSubmitter::GetActor(bool global, const std::string &actor_name) const {
+  auto &core_worker = CoreWorkerProcess::GetCoreWorker();
+  auto full_actor_name = GetFullName(global, actor_name);
+  auto pair = core_worker.GetNamedActorHandle(full_actor_name, "");
+  if (!pair.second.ok()) {
+    RAY_LOG(WARNING) << pair.second.message();
+    return ActorID::Nil();
+  }
+
+  std::string actor_id;
```
Comment (jovany-wang) [inline cpp/src/ray/runtime/task/native_task_submitter.cc:102]: Remove unused `actor_id`.

**Review by jovany-wang — APPROVED** [review]
LGTM.

**Comment by jovany-wang** [comment]
This PR only changes cpp worker.
Failure dask:dask_ray_shuffle_optimization_client_mode is unrelated.

**Pull request merged**